doc/dev/python_scheduling/autotune_inner_reduction.py

```diff
@@ -0,0 +1,395 @@
+    ScriptConfiguration,
+    collect_data,
+    separate_data,
+    test_model_rmse,
+    test_model,
+    ceil_div,
+    floor_div,
+)
+
+
+# ================================ Description ================================
+
+# This script defines four inner reduction fusions:
+#
+# 1. Inner Sum
+#    y = sum(x, dim=-1)
+#
+# 2. Add Sum
+#    z = sum(x1 + x2 + x3 + x4, dim=-1)
+#
+# 3. Tanh Sum
+#    y = sum(tanh(x), dim=-1)
+#
+# 4. Exp Sum
+#    z = sum(exp(x), dim=-1)
+#
+# Script Sequence:
+#
+# 1. Define a nvfuser fusion and its pytorch eager mode reference.
+#
+# 2. Profile the CUDA kernel performance by iterating over a set of input
+# arguments and scheduler configurations.
+#
+# 3. Train a regression model to predict the desired performance metric given
+# some input arguments and a scheduler configuration.
+#
+# 4. Measure the performance of the regression model.
+#  - Calculate RMSE of predicted and actual performance on test set.
+#  - Find the configuration with the best performance using regression model.
+#    Then, compare against the heuristic configuration selected by nvfuser.
+#  - For a specific batch size, gather performance across a range of hidden
+#    sizes. Calculate performance for best predicted and nvfuser
+#    configurations. Plot a chart comparing performance using matplotlib.
+#
+# The selected performance metric is effective_bandwidth_gbs. The empirical
+# scheduler selects the configuration that has the highest predicted
+# effective_bandwidth_gbs.
+
+# =============================================================================
+
+
+class AutotuneInnerReduction:
+    class FUSION(Enum):
+        INNER_SUM = 1
+        ADD_SUM = 2
+        TANH_SUM = 3
+        EXP_SUM = 4
+
+    @dataclass(unsafe_hash=True)
+    class InnerReductionConfiguration:
+        vectorize_factor: int = 1
+        unroll_factor: int = 1
+        godim: int = -1
+        grdim: int = -1
+        bdimx: int = -1
+        bdimy: int = -1
+
+    def __init__(self, selected_fusion):
+        self.selected_fusion = selected_fusion
+
+        # gpu device properties are defined globally
+        assert torch.cuda.is_available()
+        self.gpu_properties = torch.cuda.get_device_properties(device=0)
+
+    def __repr__(self):
+        return f"inner_reduction_{self.selected_fusion.name}"
+
+    def convert_to_inner_reduction_params(self, scheduler_config, reduction_params):
+        warp_size = 32
+        max_number_of_threads_cta = 1024
+        grid_x_limit = 2147483647
+        grid_y_limit = 65535
+
+        reduction_params.schedule_3D = False
+        reduction_params.fastest_dim = True
+        reduction_params.cross_block_inner_reduction = True
+        reduction_params.block_dim_inner_reduction = ParallelType.block_x
+        reduction_params.cross_grid_inner_reduction = scheduler_config.grdim > 1
+        reduction_params.multiple_reds_per_blk = scheduler_config.bdimy > 1
+        reduction_params.pad_inner_reduction_to_warp = (
```

Comment:
should be `bdimx >= warp && bdimx * bdimy <= max threads per block && bdimx % warp ==0`

Reply:
In the cpp code, I see:
```cpp
bool pad_bdimx = bdimx > 16 && bdimx * bdimy < threads_per_block;
```
and
```cpp
  bool pad_bdimx = bdimx > 16 &&
      bdimx * bdimy <
          (int64_t)at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;
```